File: solutions/332/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Fri Nov  5 09:20:59 2021
> Description:   
 ************************************************************************/
#include <iostream> 
#include <unordered_map> 
#include <map> 
#include <utility> 
#include <string> 
#include <gtest/gtest.h>

using namespace std; 

class Solution {
private:
    // unordered_map<from, map<to, flight no>>
    unordered_map<string, map<string, int>> targets;
    vector<string> result;
public:
    bool backtracking(vector<vector<string>>& tickets) {
        // base case
        if (result.size() == tickets.size()+1) {
            return true;
        }
        // const string since key in map is not changable
        for (pair<const string, int>& target: targets[result[result.size()-1]]) {
            if (target.second > 0) {
                result.push_back(target.first);
                target.second--;
                // note, need to return here if true
                if (backtracking(tickets)) return true;
                result.pop_back();
                target.second++;
            }
        }
        return false;
    }
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        targets.clear();
        for (const vector<string>& vec : tickets) {
            targets[vec[0]][vec[1]]++;
        }
        result.push_back("JFK");
        bool success = backtracking(tickets);
        return result;
    }
};

TEST(Test332, SimpleTest) {
    vector<vector<string>> tickets; 
    vector<string> ticket1{
        "MUC","LHR"
    };
    vector<string> ticket2{
        "JFK","MUC"
    };
    vector<string> ticket3{
        "SFO","SJC"
    };
    vector<string> ticket4{
        "LHR","SFO"
    };
    tickets.push_back(ticket1);
    tickets.push_back(ticket2);
    tickets.push_back(ticket3);
    tickets.push_back(ticket4);
    Solution s; 
    vector<string> result = s.findItinerary(tickets);
    cout << "itineray ";
    for (auto& r : result) 
        cout << r << " ";
    cout << endl; 
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/332/main.py
from typing import List 
from collections import defaultdict 

class Solution:
    def __init__(self):
        self.result = []
        self.targets = defaultdict(list)
    def backtracking(self, tickets):
        # base case
        if len(self.result) == len(tickets)+1:
            return True
        start = self.result[-1]
        # NOTE, need to sort here
        self.targets[start].sort()
        for _ in self.targets[start]:
            dest = self.targets[start].pop(0)
            self.result.append(dest)
            if self.backtracking(tickets):
                return True
            self.result.pop()
            self.targets[start].append(dest)
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        # targets {'JFK': ['SFO', 'ATL'], 'SFO': ['ATL'], 'ATL': ['JFK', 'SFO']})
        for item in tickets:
            self.targets[item[0]].append(item[1])
        self.result.append("JFK")
        self.backtracking(tickets)
        return self.result 

if __name__ == "__main__":
    s = Solution()
    tickets = [["MUC","LHR"],["JFK","MUC"],["SFO","SJC"],["LHR","SFO"]]
    print(s.findItinerary(tickets))
